### PR TITLE
SDK-306 [Node.js SDK] add Schema 2.4 support

### PR DESCRIPTION
**Pull request by igor-alekseev-dev**
Pull Request Checklist
======================

Please read this for our PR Culture:
------------------------------------
https://incountry.atlassian.net/wiki/spaces/ED/pages/46858869/Pull+requests+culture+Draft

Work relates to:
----------------
Link(s) to Jira Task/User Story/Bug/Epic:
- [SDK-306](https://incountry.atlassian.net/browse/SDK-306)

Code Quality (Before submitting PR)
------------
- [x] Ensure code compiles and passes Sonar Lint.
- [x] Do your best to ensure integration tests pass. If they don't then...
- [x] Remove, ignore, or mark pending any brittle tests.
- [x] Remove commented out code.

Target Release
--------------
- [ ] Please put the target release vehicle this needs to be bundled with.
- [ ] Did you increment the version in package.json? What is the new version?

### CODE REVIEW

File: src/auth-client.ts [hunk 1]
```diff
@@ -19,7 +19,7 @@ const DEFAULT_HEADERS = {
 const ERROR_RESPONSES: Record<string, string> = {
   invalid_request: 'The request is missing a required parameter, includes an invalid parameter value, includes a parameter more than once, or is otherwise malformed.',
   invalid_client: 'Client authentication failed (e.g., unknown client, no client authentication included, or unsupported authentication method).',
-  invalid_grant: 'The provided authorization grant (e.g., authorization code, resource owner credentials) or refresh token is invalid, expired, revoked, does not match the redirection URI used in the authorization request, or was issued to another client.',
+  invalid_grant: 'The provided authorization grant (e.g., authorization code, resource owner credentials) or refresh token is invalid, expires, revoked, does not match the redirection URI used in the authorization request, or was issued to another client.',
```
Comment (RommeroInCountry) [inline src/auth-client.ts:22]: expired :)

**Review by RommeroInCountry — CHANGES_REQUESTED** [review]

File: README.md [hunk 2]
```diff
@@ -319,6 +322,14 @@ rangeKey9
 rangeKey10
 ```
 
+##### Date fields, plain:
+```typescript
+expiresAt
+```
+
+**WARNING** The data with `expiresAt` will be vanished when time comes ))).
```
Comment (RommeroInCountry) [inline README.md:330]: Records with non-null `expiresAt` value will be automatically deleted upon reaching the specified date.

File: src/validation/user-input/find-options.ts [hunk 3]
```diff
@@ -131,7 +139,7 @@ const sortItemValidate = (i: { [key: string]: unknown }): Either<string, Sorting
     return left(`"${Object.keys(i)[0]}" is not allowed for sorting. Check documentation https://github.com/incountry/sdk-nodejs#find-records`);
   }
 
-  if (!SortingDirection.is(Object.values(i)[0])) {
+  if (!SortingDirectionIO.is(Object.values(i)[0])) {
     return left('Only "asc" and "desc" is allowed as sorting direction');
```
Comment (RommeroInCountry) [inline src/validation/user-input/find-options.ts:143]: direction => order (sort order)

File: src/validation/user-input/storage-record-data-array.ts [hunk 4]
```diff
@@ -0,0 +1,43 @@
+import * as t from 'io-ts';
+import {
+  left, right, Either, either,
+} from 'fp-ts/lib/Either';
+import { getStorageRecordDataIO, StorageRecordData } from './storage-record-data';
+import { Codec, chainValidate } from '../utils';
+
+const MAX_RECORDS_IN_BATCH = 100;
+const STORAGE_RECORD_DATA_ARRAY_EMPTY_ERROR = 'You must pass non-empty array of records';
+const STORAGE_RECORD_DATA_ARRAY_TOO_BIG_ERROR = `You must pass array of not more than ${MAX_RECORDS_IN_BATCH} records`;
```
Comment (RommeroInCountry) [inline src/validation/user-input/storage-record-data-array.ts:10]: You mast pass an array not larger than X records

**Review by RommeroInCountry — DISMISSED** [review]

File: src/validation/user-input/find-filter.ts [hunk 5]
```diff
@@ -0,0 +1,220 @@
...
+  'key16',
+  'key17',
+  'key18',
+  'key19',
+  'key20',
+];
+const SEARCH_FIELD_MIN_LENGTH = 3;
+const SEARCH_FIELD_MAX_LENGTH = 200;
+
+const SEARCH_FIELD_LENGTH_ERROR_MESSAGE = `filter.${SEARCH_FIELD} should be at least ${SEARCH_FIELD_MIN_LENGTH} but not more than ${SEARCH_FIELD_MAX_LENGTH}`;
```
Comment (artemgolozubov) [inline src/validation/user-input/find-filter.ts:33]:
```suggestion
const SEARCH_FIELD_LENGTH_ERROR_MESSAGE = `filter.${SEARCH_FIELD} length should be between ${SEARCH_FIELD_MIN_LENGTH} and ${SEARCH_FIELD_MAX_LENGTH} chars`;
```

File: src/validation/user-input/storage-record-data-array.ts [hunk 6]
```diff
@@ -0,0 +1,43 @@
+import * as t from 'io-ts';
+import {
+  left, right, Either, either,
+} from 'fp-ts/lib/Either';
+import { getStorageRecordDataIO, StorageRecordData } from './storage-record-data';
+import { Codec, chainValidate } from '../utils';
+
+const MAX_RECORDS_IN_BATCH = 100;
+const STORAGE_RECORD_DATA_ARRAY_EMPTY_ERROR = 'You must pass non-empty array of records';
+const STORAGE_RECORD_DATA_ARRAY_TOO_BIG_ERROR = `You mast pass an array not larger than ${MAX_RECORDS_IN_BATCH} records`;
```
Comment (artemgolozubov) [inline src/validation/user-input/storage-record-data-array.ts:10]:
```suggestion
const STORAGE_RECORD_DATA_ARRAY_TOO_BIG_ERROR = `You must pass an array not larger than ${MAX_RECORDS_IN_BATCH} records`;
```

File: tests/unit/api-client-test.ts [hunk 7]
```diff
@@ -372,12 +375,13 @@ describe('ApiClient', () => {
...
           const record = {
-            ...EMPTY_API_RECORD,
+            ...EMPTY_API_RESPONSE_RECORD,
             record_key,
           };
           const popAPI = nockPopApi(POPAPI_HOST).read(COUNTRY, record_key).reply(200, record);
           const result = await apiClient.read(COUNTRY, record_key);
-          expect(result).to.deep.equal(record);
+          expect(result).to.deep.equal(toApiRecord(record));
+          //
```
Comment (artemgolozubov) [inline tests/unit/api-client-test.ts:384]: Do we need this empty comment?

**Review by artemgolozubov — APPROVED** [review]
LGTM in general, but fix the typos please

File: src/validation/user-input/find-filter.ts [hunk 8]
```diff
@@ -30,7 +30,7 @@ const EXCLUDED_KEYS_WHEN_SEARCHING = [
 const SEARCH_FIELD_MIN_LENGTH = 3;
 const SEARCH_FIELD_MAX_LENGTH = 200;
 
-const SEARCH_FIELD_LENGTH_ERROR_MESSAGE = `filter.${SEARCH_FIELD} should be at least ${SEARCH_FIELD_MIN_LENGTH} but not more than ${SEARCH_FIELD_MAX_LENGTH}`;
+const SEARCH_FIELD_LENGTH_ERROR_MESSAGE = `filter.${SEARCH_FIELD} should be length should be between ${SEARCH_FIELD_MIN_LENGTH} and ${SEARCH_FIELD_MAX_LENGTH} chars`;
```
Comment (RommeroInCountry) [inline src/validation/user-input/find-filter.ts:33]: still a typo %)

**Review by RommeroInCountry — CHANGES_REQUESTED** [review]

**Review by artemgolozubov — DISMISSED** [review]
LGTM

**Review by artemgolozubov — APPROVED** [review]
LGTM

**Review by RommeroInCountry — APPROVED** [review]

**Pull request merged**